src/duckling/canvas/drawing/entity-drawer.service.ts

```diff
@@ -70,49 +60,79 @@ export class EntityDrawerService extends BaseAttributeService<AttributeDrawer<At
         return null;
     }
 
+    paintAttribute(key : AttributeKey, entity : Entity, graphics : Graphics) {
+        let drawer = this.getImplementation(key);
+        if (drawer && drawer.paint) {
+            drawer.paint(entity[key], this._entityPosition.getPosition(entity), graphics);
+        }
+    }
 
-    /**
-     * Get a DisplayObject for the entity.
-     * @param entity Entity that needs to be drawn.
-     * @return A DisplayObject for the entity.
-     */
-    drawEntity(entity : Entity) : DrawnConstruct[] {
+    getEntityDrawable(entity : Entity) : DrawnConstruct[] {
         let drawnConstructs : DrawnConstruct[] = [];
         for (let key in entity) {
             if (!this._layers.isEntityAttributeVisible(entity, key)) {
                 continue;
             }
-            let drawableConstruct = this.drawAttribute(key, entity);
+            let drawableConstruct = this.getAttributeDrawable(key, entity);
             if (drawableConstruct) {
                 drawnConstructs.push(drawableConstruct);
             }
         }
         return drawnConstructs;
     }
 
+    paintEntity(entity : Entity, graphics : Graphics) {
+        if (!this._layers.isEntityVisible(entity)) {
+            return;
+        }
+
+        for (let key in entity) {
+            this.paintAttribute(key, entity, graphics);
+        }
+    }
+
+    drawEntitySystem(entitySystem : EntitySystem, graphics? : Graphics) : DrawnConstruct[] {
+        let sortedEntities = this._renderPriority.sortEntities(entitySystem);
+        if (graphics) {
+            this._paintEntitySystem(sortedEntities, graphics);
+        }
+        return this._getEntitySystemDrawables(sortedEntities);
+    }
+
+    private _getEntitySystemDrawables(entities : TaggedEntity[]) : DrawnConstruct[] {
+        let drawnConstructs : DrawnConstruct[] = [];
+        let newCache : EntityCache = {};
```

Comment:
I think we should have the cache held by the MapDrawer and this should expose a InvalidateDrawableCache observer that fires an event when something causes the drawable cache to be stale.